File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = float('inf')
    arr = [[] for _ in range(n+1)]
    
    for w1, w2 in wires:
        arr[w1].append(w2)
        arr[w2].append(w1)
        
    def bfs(s, cut1, cut2):
        Q = deque()
        Q.append(s)
        visited = set([s])
        
        while Q:
            c = Q.popleft()
            
            for next in arr[c]:
                if (cut1 == c and cut2 == next) or (cut2 == c and cut1 == next):
                    continue
                
                if next not in visited:
                    visited.add(next)
                    Q.append(next)
                    
        return len(visited)
        
    for w1, w2 in wires:
        cnt_subtree = bfs(w1, w1, w2)
        dif = abs(cnt_subtree - (n - cnt_subtree))
        answer = min(dif, answer)
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    let answer = Infinity;
    
    // arr = [[] for _ in range(n+1)]
    const graph = Array.from({ length: n + 1}, () => []);
    
    wires.forEach(([w1, w2]) => {
        graph[w1].push(w2);
        graph[w2].push(w1);
    });
    
    function bfs(start, cut1, cut2) {
        const queue = [start];
        const visited = new Set([start]);
        
        while (queue.length > 0) {
            const current = queue.shift();
            
            for (const next of graph[current]) {
                if ((cut1 === current && cut2 === next) || (cut2 === current && cut1 === next)) {
                    continue;
                }
                
                if (!visited.has(next)) {
                    visited.add(next);
                    queue.push(next);
                }
            }
        }
        return visited.size;
    }
    
    wires.forEach(([w1, w2]) => {
        const cntSubtree = bfs(w1, w1, w2);
        const dif = Math.abs(cntSubtree - (n - cntSubtree));
        answer = Math.min(answer, dif);
    });
    
    return answer;
}